### ISSUE TITLE
Easy way to get content of an element?

### ISSUE THREAD

**Issue by bee-git** [atroche/rust-headless-chrome#73]
Am I missing something, or is it actually very difficult with the current code to just get the content of a page element once you have done a `wait_for_element()`?

To be honest, I haven't been able to do it yet. I've "waited for an element", then done `get_description()` and tried to get the elements value through both `.node_value` and plain `.value`, but the content comes up emtpy everytime I've tried (and I've tried with different web pages, include pages with no dynamically loaded content).

I've been using the code in the example "query_wikipedia.rs" as a guide for trying to print the contents of an element, although I have not used the `find` method since I'm looking for the content of a single element.

Thanks!

**Comment by atroche** [comment]
No, you're right, it is hard at the moment! Just to clarify, when you mean content, do you mean https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent ?

**Comment by atroche** [comment]
(I can get a PR up pretty quickly, I reckon, if you let me know what you're after. Is it something like `element.text()`?

**Comment by bee-git** [comment]
Yes, it's the `textContent` that I'm looking for -- the html within the element that is then rendered by the browser.

Once I figure out how to do it, I'll be happy to contribute a related example. However, at this point, I can't even get it to work in what should be relatively simple cases.

Thanks for the quick reply and for all your work on this valuable library.

**Comment by atroche** [comment]
@bee-git so for this HTML:

```html
<div id="root">
    Some text <b>with part of it in <a href="#">nested elements</a></b>
</div>
```

Would you want a method to return (when called on the `div#root`) node?

```
"Some text with part of it in nested elements"
```

I think the most confusing part of doing that (with the existing, quite low-level `get_description()` is that the text inside (e.g.) a div node is actually considered a separate node.

**Comment by atroche** [comment]
P.S. you can kind of work around it by doing this at the moment:

```rust
    let root_div: Element = tab.wait_for_element("#root")?;
    let desc = root_div.get_description();
    let text = root_div.call_js_fn("function() { return this.textContent;}");
```

**Comment by bee-git** [comment]
What I personally am after is the HTML, not the text that would result from it. So I think it would be more like `element.html()` .

So, to use your example, what I'm really looking to have returned is: 

`Some text <b>with part of it in <a href="#">nested elements</a></b>`

As I said, I tried using `get_description()` and then working with `.node_value` and just `.value`, all based on the result of a `wait_for_element()`. In all cases, I came up with an empty result. But this may be explained by what you just wrote about what is inside a div node actually being a separate node.

**Comment by atroche** [comment]
From which you get back this:

```rust
[tests/simple.rs:156] text = 
    RemoteObject {
        object_type: "string",
        subtype: None,
        description: None,
        class_name: None,
        value: Some(
            String(
                "\n    Some text with part of it in nested elements\n"
            )
        ),
        unserializable_value: None,
        preview: None
    }
```

**Comment by atroche** [comment]
Ah, for the HTML, you could do:

```rust
    let root_div: Element = tab.wait_for_element("#root")?;
    let html = root_div.call_js_fn("function() { return this.innerHTML;}")?.value.unwrap();
```

```rust
[tests/simple.rs:156] html = String(
    "\n    Some text <b>with part of it in <a href=\"#\">nested elements</a></b>\n"
)
```

**Comment by bee-git** [comment]
Thanks. I'll work with this and see if it lets me get what I'm looking for.

I had not picked up on the fact that you can run arbitrary javascript with the library. Obviously, that opens up a lot of possibilities, although using Javascript to get things you want from a page seems to defeat the purpose of using a Rust library!

**Comment by atroche** [comment]
Yeah, we'll definitely want those things behind methods on Element :)

Thanks for being patient.

**Comment by bee-git** [comment]
Just wanted to let you know that working with Javascript to pull the HTML I want has been working very well.

The one issue I've encountered, though, which may be related to how you've implemented this crate is that HTML coming into Rust may not be serializing entirely correctly, at least in some cases.

I've been testing on a page that is UTF-8 encoded and has some special characters occasionally. The example that is standing out is the "≥" character. When I use the `call_js_fn` approach you suggested above and then dump the contents of the object for inspection, the instances of "≥" in the String appear as "â‰¥". 

This does not appear to be a problem in the source page itself. If, after the page has fully loaded in a normal browser, I save the HTML with right clicking and selecting "Save as", the instances of "≥" appear correctly in the resulting HTML file. Ditto for how they appear on the page itself.

Here is the page I'm using for testing purposes, in case you're interested:

[https://tct.confex.com/tct/2019/meetingapp.cgi/Paper/13530](https://tct.confex.com/tct/2019/meetingapp.cgi/Paper/13530)

Thanks again for your suggestions earlier in this thread and your work on this crate.

**Comment by atroche** [comment]
Cool, thanks for the heads up, I'll definitely take a look, but not until Monday at the earliest unfortunately :)

**Comment by bee-git** [comment]
That's fine. It's not urgent. It's mainly a heads up to you that something may be going with the serialization that isn't entirely correct. If I find a solution, I'll let you know.

Have a great weekend!

**Comment by bee-git** [comment]
I've spent a lot of time looking into the UTF-8 character encoding issue I mentioned above. It turns out that there is NOT an issue.

The reason there appeared to be an issue is because of the peculiarities of my work setup. I do my Rust work in a CentOS Virtualbox client. When I want to check the HTML output by my Rust programs, I transfer the HTML file over to my Windows system and view it there. Or I use a Windows application (WinSCP) to view the HTML file on the Linux client. 

The Windows applications were not smart enough to interpret the UTF-8 encoding of special characters in the HTML files generated by my Rust programs. Those characters are displayed just as they should be when I view them within the Linux system.

Long story short: headless-chrome is doing what it should be doing. There is no encoding issue as far as I can tell.

**Comment by atroche** [comment]
Thanks for checking in! I'm very glad to hear that I won't have to look into weird encoding issues in the near future (touch wood!).
